mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+ *
+ * </pre>
+ *
+ * If the document is modified by another thread, that update will not
+ * appear as an update in the result. But, we do not know which
+ * documents were failed to update this way, only the number of
+ * documents actually updated. So, we read the database again to see
+ * which docvers have been changed. Another thread may come an modify
+ * the documents before this step but after the modifications are
+ * applied, meaning we may find out more documents have been modified
+ * than initially known. Here the question is which documents have our
+ * modifications, and which don't. We would like to filter out the
+ * documents that don't have our modifications, and mark them as
+ * concurrent update failures, because any update performed after our
+ * update is valid.
+ *
+ * This is important for incremental updates. Example:
+ * <pre>
+ *   concurrent update failure:
+ *      thread1:                        thread2:
+ *      read doc, ver=1, field=1
+ *                                read doc, ver=1, field=1
+ *      set field=2
+ *      submit update with ver=2
+ *                                set field=2
+ *                                submit update with ver=2
+ *                                perform db update
+ *      update fails, ver!=1
+ * </pre>
+ *
+ * In the above scenario, the increment operation for thread1 should
+ * fail, because thread2 modified the document after thread1 read
+ * it. This is returned as one missing document in the write result of
+ * thread 1. At this point, thread1 doesn't know which document
+ * failed, so it reads the db to find out:
+ * <pre>
+ *   thread1:                            thread2:
+ *                                     modify another doc, ver=2, set ver=3
+ *   read all docs with ver=2
+ *   
+ * </pre>
+ *
+ * At this point, thread1 sees two documents missing, instead of
+ * one. The first document doesn't have thread1 updates, the second
+ * document has thread1 updates, but thread2 applied its own updates
+ * before thread1 can find out which document was modified. Returning
+ * the second document as failed might result in the client retrying
+ * the update, applying the changes twice,
+ *
+ * To prevent this scenario, we will keep the document versions in an
+ * array, so we'll know which updates are applied to the document. To
+ * prevent unbounded array growth, we will truncate the array using
+ * the docver timestamp.
+ *
+ */
+
+
+
+public class MongoSafeUpdateProtocol {
+
+    private static final Logger LOGGER=LoggerFactory.getLogger(MongoSafeUpdateProtocol.class);
+    
+    public static final String DOCVER="docver";
+
+    public static final String DOCVER_FLD=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER;
+    public static final String DOCVER_FLD0=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER+".0";
+
+    private static final long TOO_OLD=10l*60l*1000l; // Any docver older than 10 minutes is to old
```

Comment:
Arbitrary time that's not too short or too long. Maybe 10 minutes is too
long? 1 minute?

On Fri, Aug 26, 2016 at 9:47 AM, Brandon Vulaj <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/
> MongoSafeUpdateProtocol.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303#discussion_r76441140>
> :
>
> > + * the docver timestamp.
> > + *
> > + */
> > +
> > +
> > +
> > +public class MongoSafeUpdateProtocol {
> > +
> > +    private static final Logger LOGGER=LoggerFactory.getLogger(MongoSafeUpdateProtocol.class);
> > +
> > +    public static final String DOCVER="docver";
> > +
> > +    public static final String DOCVER_FLD=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER;
> > +    public static final String DOCVER_FLD0=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER+".0";
> > +
> > +    private static final long TOO_OLD=10l*60l*1000l; // Any docver older than 10 minutes is to old
>
> Why 10 minutes? Just a safe amount of time?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303/files/6e592caaaf4c79fd821914fdb9640896fa145ab0#r76441140>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADgDDb5AnIf69NF3j_A6zKUemHM3x6Ysks5qjwqcgaJpZM4JtbtZ>
> .
>